tools/LLVMOptimizer.h

```diff
@@ -0,0 +1,52 @@
+#pragma once
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/ADT/Triple.h"
+#include "llvm/Analysis/MemorySSA.h"
+#include "llvm/Analysis/TargetLibraryInfo.h"
+#include "llvm/Bitcode/BitcodeReader.h"
+#include "llvm/IR/CFG.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/LLVMContext.h"
+#include "llvm/IR/LegacyPassManager.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/Verifier.h"
+#include "llvm/IRReader/IRReader.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/FormatVariadic.h"
+#include "llvm/Support/PrettyStackTrace.h"
+#include "llvm/Support/Signals.h"
+#include "llvm/Support/SourceMgr.h"
+#include "llvm/Transforms/IPO/PassManagerBuilder.h"
+#include "llvm/Transforms/Scalar/LICM.h"
+#include "llvm/Transforms/Scalar/NewGVN.h"
+
+#include <cassert>
+#include <climits>
+#include <ctime>
+#include <memory>
+#include <random>
+#include <string>
+#include <vector>
+
+class LLVMOptimizer {
+  std::string optArgs;
+  llvm::LoopAnalysisManager LAM;
+  llvm::FunctionAnalysisManager FAM;
+  llvm::CGSCCAnalysisManager CGAM;
+  llvm::ModuleAnalysisManager MAM;
+  llvm::PassBuilder PB;
+
+  llvm::FunctionPassManager FPM;
+  llvm::ModulePassManager MPM;
+
+public:
+  LLVMOptimizer(std::string optArgs);
```

Comment:
do we need all this OO interface or is it an overkill? The current usage creates an object, calls 1 function, an throws the object away. So the OO seems overkill.

Reply:
Yes.. I think you are right. This class won't have too many extensions and two functions are enough.

Reply:
How about static functions inside a class? or just functions

Reply:
what about a single function: `void optimize_module(llvm::Module *M)`?
(plus an `optimize_function` if you need it for your uses)

Reply:
OK, I can do that. The return value is just for convenience.

Reply:
fixed